### PR TITLE
[clang][ItaniumMangle] Mangle friend function templates with a constr…

### LINKED ISSUE #110088: [abi] Declaring class of a member-like friend method is not mangled.
The code sample

```cpp
template <class>
concept True = true;

namespace test2 {
 template<typename T> struct A {
    template<typename U = void>
    friend void g(...) requires True<T> && True<U> {}
 };

 void call() {
    A<int> ai;
    A<bool> ab;
    g(ai);
    g(ab);
 }
}
```

does not compile with the following error 

```
Output of x86-64 clang (trunk) (Compiler #1)
<source>:10:17: error: definition with same mangled name '_ZN5test2F1gIvEEvzQaa4TrueIT_E4TrueITL0__E' as another definition
   10 |     friend void g(...) requires True<T> && True<U> {}
      |                 ^
<source>:10:17: note: previous definition is here
1 error generated.
Compiler returned: 1
```

The reason appears to be that the enclosing class template specialization `A<int>` (or `A<bool>`, respectively) is not part of the mangled name `_ZN5test2F1gIvEEvzQaa4TrueIT_E4TrueITL0__E`, resulting in `g(ai)` and `g(ab)` mangled as the same string.

As per the discussion in https://github.com/itanium-cxx-abi/cxx-abi/issues/24#issuecomment-934493006, the code above is valid and the solution is to include the declaring class in the mangling. From @rjmccall on the aforementioned issue:

```
The only viable implementation path I can see, given that nothing in the standard prevents these friend declarations from otherwise matching perfectly in signatures and requires-clauses, is to mangle the declaring class for these friends.
```

Note 1: gcc can compile the above snippet by including the declaring class in the mangling, which results in the mangled string `_ZN5test21AIiEF1gIvEEvzQaa4TrueIT_E4TrueITL0__E` for `g(ai)` (Note `AIiE` mangles the declaring class),  and `_ZN5test21AIbEF1gIvEEvzQaa4TrueIT_E4TrueITL0__E`  for `g(ab)` (Note `AIbE` mangles the declaring class).

Note 2: this issue stands in the way of implementing the demangling of template parameters in constraints (https://github.com/llvm/llvm-project/blob/main/libcxxabi/src/demangle/ItaniumDemangle.h#L5795).
`T_` in `_ZN5test21AIiEF1gIvEEvzQaa4TrueIT_E4TrueITL0__E` is impossible to demangle, since it is a substitution for the template argument of class A that is not part of the mangling.

### PR DESCRIPTION
**Pull request by VitaNuo**
…aint that depends on a template parameter from an enclosing template as members of the enclosing class.

Such function templates should be considered member-like constrained friends per [temp.friend]p9 and https://github.com/itanium-cxx-abi/cxx-abi/issues/24#issuecomment-934977198).

### CODE REVIEW

**Review by usx95 — COMMENTED** [review]
Thanks for catching and fixing this.

Please also add release notes in clang/docs/ReleaseNotes.rst

File: clang/test/CodeGenCXX/mangle-concept.cpp [hunk 1]
```diff
@@ -58,19 +58,19 @@ namespace test2 {
     // CHECK: call {{.*}}@_ZN5test21AIiEF1fEzQ4TrueIT_E(
     // CLANG17: call {{.*}}@_ZN5test21fEz(
     f(ai);
-    // CHECK: call {{.*}}@_ZN5test2F1gIvEEvzQaa4TrueIT_E4TrueITL0__E(
+    // CHECK: call {{.*}}@_ZN5test21AIiEF1gIvEEvzQaa4TrueIT_E4TrueITL0__E(
```
[the remark this answers is not in the document]
Comment (VitaNuo) [reply clang/test/CodeGenCXX/mangle-concept.cpp:62]: Sure.

File: clang/lib/AST/ItaniumMangle.cpp [hunk 2]
```diff
@@ -2270,6 +2270,14 @@ void CXXNameMangler::mangleTemplatePrefix(GlobalDecl GD,
     mangleTemplateParameter(TTP->getDepth(), TTP->getIndex());
   } else {
     const DeclContext *DC = Context.getEffectiveDeclContext(ND);
+    if (const auto *FD = dyn_cast<FunctionTemplateDecl>(GD.getDecl())) {
```
[the remark this answers is not in the document]
Comment (VitaNuo) [reply clang/lib/AST/ItaniumMangle.cpp]: Makes sense, thank you.

File: clang/lib/AST/ItaniumMangle.cpp [hunk 3]
```diff
@@ -704,6 +704,15 @@ ItaniumMangleContextImpl::getEffectiveDeclContext(const Decl *D) {
       return D->getLexicalDeclContext()->getRedeclContext();
   }
 
+  if (const auto *FTD = dyn_cast<FunctionTemplateDecl>(D)) {
+    // Member-like constrained friends are mangled as if they were members of
```
Comment (usx95) [inline clang/lib/AST/ItaniumMangle.cpp:708]: Let's reuse the function decl above.

We could use the `Decl::getAsFunction()` to do both of these checks for us.
```cpp
if (const auto *FD = D->getAsFunction())
```

**Review by usx95 — APPROVED** [review]
Thanks. LGTM.

Left a comment to simplify the change.

File: clang/lib/AST/ItaniumMangle.cpp (hunk 3 above)
Comment (VitaNuo) [reply clang/lib/AST/ItaniumMangle.cpp]: Great idea!

**Comment by zygoloid** [comment]
Should this have `-fclang-abi-compat` support?

**Comment by hokein** [comment]
> Should this have -fclang-abi-compat support?

(It seems like we should, although I don’t have direct experience with it.)

This is a bug fix for https://github.com/llvm/llvm-project/commit/4b163e343cfa54c8d55c9da73c70d58f55ea9df2, which is released in Clang 18 and Clang 19. Therefore, we should likely add -fclang-abi-compat for `Ver19`?

**Pull request merged**